### DISCUSSION TITLE
Horovod Model Determinism Issues

### DISCUSSION THREAD

**Discussion by ghost** [horovod/horovod#2779] [Q&A]
Horovod versions used: 0.20.3/ 0.21.1

Clusters Used:
- Databricks  1 worker 2 gpu, i.e np=2
- Databricks  3 workers 4 gpus each i.e. np=12

Issues:
- Horovod Model Weights and Training Logs are slightly different for every run

With version 0.20.3, I had to manually set `os.environ['HOROVOD_FUSION_THRESHOLD'] = '0'`, but it worked for small dataset(200-300k records) and small number of workers. 

The moment I increased the dataset to say a few million records or increased the number of workers, the model weights were differing every run in their tenth/hundredth place. Also, the losses were different.

I also had a similar issue with version 0.21.1 by setting `num_groups` parameter to some number (tried from 1 to `np`) while calling DistributedOptimizer. Worked well on the small dataset and np=2 but not on a larger dataset with a greater number of workers.

Any suggestions regarding how to fix the Horovod Model Determinism issue are welcome.

**Comment by romerojosh** [comment]
@SRITANU Along with setting `num_groups`, can you also set `HOROVOD_DISABLE_GROUP_FUSION=1` and see if that helps?

**Comment by ghost** [comment]
@romerojosh Apologies for my delayed response. I found one of your [comments](https://github.com/horovod/horovod/issues/2616#issuecomment-765622844) related to this I set `num_groups` to 1 and :
```
os.environ['HOROVOD_FUSION_THRESHOLD'] = '1024'
os.environ['HOROVOD_DISABLE_GROUP_FUSION'] = '1'
```
And ran the training with `np=4` for 10 epochs. Still the same issue. The model weights were differing every run in their tenth/hundredth place.
